src/Simulation/Simulators/CommonNativeSimulator/Dump.cs

```diff
@@ -96,5 +92,393 @@ public QSimDumpRegister(CommonNativeSimulator m) : base(m)
+            /// <returns>true if dumping should continue, false to stop dumping.</returns>
+            public abstract bool Callback(uint idx, double real, double img);
+
+            /// <summary>
+            /// The Simulator being reported.
+            /// </summary>
+            public CommonNativeSimulator Simulator { get; }
+
+            /// <summary>
+            /// Entry method to get the dump of the wave function.
+            /// </summary>
+            public virtual bool Dump(IQArray<Qubit>? qubits = null)
+            {
+                if (qubits == null)
+                {
+                    this.Simulator.sim_Dump(Callback);
+                    return true;
+                }
+                else
+                {
+                    var ids = qubits.GetIds();
+                    return this.Simulator.sim_DumpQubits((uint)ids.Length, ids, Callback);
+                }
+            }
+        }
+
+        /// <summary>
+        ///     A state dumper that encodes dumped states into displayable
+        ///     objects.
+        /// </summary>
+        public class DisplayableStateDumper : StateDumper
+        {
+            private long _count = -1;
+            private Complex[]? _data = null;
+
+            /// <summary>
+            /// A method to call to output a string representation.
+            /// </summary>
+            public virtual Action<string>? FileWriter { get; }
+
+            /// <summary>
+            ///     Constructs a new display dumper for a given simulator.
+            /// </summary>
+            public DisplayableStateDumper(CommonNativeSimulator sim, Action<string>? fileWriter = null) : base(sim)
+            {
+                this.FileWriter = fileWriter;
+            }
+
+            /// <summary>
+            ///     Used by the simulator to provide states when dumping.
+            ///     Not intended to be called directly.
+            /// </summary>
+            public override bool Callback(uint idx, double real, double img)
+            {
+                if (_data == null) throw new Exception("Expected data buffer to be initialized before callback, but it was null.");
+                _data[idx] = new Complex(real, img);
+                return true;
+            }
+
+            /// <summary>
+            ///     Dumps the state of a register of qubits as a displayable object.
+            /// </summary>
+            // TODO(rokuzmin): What does it return?
+            public override bool Dump(IQArray<Qubit>? qubits = null)
+            {
+                System.Diagnostics.Debug.Assert(this.Simulator.QubitManager != null, 
+                    "Internal logic error, QubitManager must be assigned");
+
+                _count = qubits == null
+                            ? this.Simulator.QubitManager.AllocatedQubitsCount
+                            : qubits.Length;
+                _data = new Complex[1 << ((int)_count)];    // If 0 qubits are allocated then the array has 
```

Comment:
No suggested action at the moment, but this will likely need to be changed to a dictionary ass well once the sparse simulator is added.

Reply:
Changed to a dictionary.